### PR TITLE
The Avengers: Infinity NaNny: JSON-serialize +/- infinity and NaN as null.

### PR DESCRIPTION
**Pull request by Shadowfiend**
Before now, we were serializing these as their Scala string equivalents. This
meant only JavaScript when evaling JSON as actual JavaScript could deserialize 
it. lift-json itself couldn't deserialize the serialized special values, in fact.

We also make this behavior customizable by providing a `RenderSettings`
parameter, `DoubleRenderer`, which is charged with taking a `Double` and
rendering it to `String`. Three variants are provided: an as-is renderer that is
identical to pre-3.1 handling of special `Double` values like `NaN`, the new
null-special-values renderer that has the behavior described above, and a
new behavior, `FailToRenderSpecialValues`, which throws an exception when
a special `Double` value is encountered rather than serializing it in any way.

Still an open question on the list:
 - [x] [Should we provide a Formats toggle for this behavior?](https://groups.google.com/d/msg/liftweb/7p9CNfClhhM/GaQy3HqDAwAJ)

This is a fix for #1780.

### CODE REVIEW

**Comment by Shadowfiend** [comment]
After thinking about this a bit more and reading a little on what jackson is doing (and [a proposal for additional customization in their renderer](https://github.com/FasterXML/jackson-databind/issues/911)), I decided to do a touch more legwork.

First off, we hook into `RenderSettings`, not `Formats`, since this happens during rendering rather than during composition to the JSON AST. Secondly, we provide a customizable `doubleRenderer`, which defaults to the new `null` rendering for special values. And finally, we add *two* variants here, one that renders doubles as-is, and one that will throw an `IllegalArgumentException` when attempting to render a special double value. This allows a little more customization for folks who may want to actually blow up when code attempts to render something their clients wouldn't expect to see.

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala [hunk 1]
```diff
@@ -881,17 +881,74 @@ object JsonAST {
...
+   * a JDouble are rendered to JSON string.
+   */
+  sealed trait DoubleRenderer extends Function1[Double,String] {
+    def apply(double: Double): String
+  }
+  /**
+   * A `DoubleRenderer` that renders special values `NaN`, `-Infinity`, and
+   * `Infinity` as-is using `toString`. This is not strictly valid JSON, but
+   * JavaScript can eval it. Other double values are also rendered the same
+   * way.
```
Comment (farmdawgnation) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: We should note that if you render JSON using this thing you may not be able to parse that some JSON back with Lift.

File: core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala [hunk 2]
```diff
@@ -34,6 +36,77 @@ object JsonPrintingSpec extends Specification  with JValueGen with ScalaCheck {
...
+        JsonAST.RenderSettings(0, doubleRenderer = JsonAST.FailToRenderSpecialValues)
+      )
+    }
+
+    "render a standard double as is" in {
+      val double = Random.nextDouble
+      render(JDouble(double)) must_== double.toString
+    }
+
+    "render positive infinity as null" in {
```
Comment (farmdawgnation) [inline core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala]: Should read something like `throw an exception if attempting to render positive infinity`

File: core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala [hunk 3]
```diff
@@ -34,6 +36,77 @@ object JsonPrintingSpec extends Specification  with JValueGen with ScalaCheck {
...
+      render(JDouble(Double.PositiveInfinity)) must throwAn[IllegalArgumentException]
+    }
+
+    "render negative infinity as null" in {
```
Comment (farmdawgnation) [inline core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala]: Should read something like `throw an exception if attempting to render negative infinity`

File: core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala [hunk 4]
```diff
@@ -34,6 +36,77 @@ object JsonPrintingSpec extends Specification  with JValueGen with ScalaCheck {
...
+      render(JDouble(Double.NegativeInfinity)) must throwAn[IllegalArgumentException]
+    }
+
+    "render NaN as null" in {
```
Comment (farmdawgnation) [inline core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala]: Should read something like `throw an exception if attempting to render NaN`

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala [hunk 5]
```diff
@@ -881,17 +881,74 @@ object JsonAST {
...
+  }
+  /**
+   * A `DoubleRenderer` that throws an `IllegalArgumentException` when the
+   * special values `NaN`, `-Infinity`, and `Infinity` are encountered. Other
+   * doubles are rendered normally using `toString`.
+   */
+  case object FailToRenderSpecialValues extends DoubleRenderer {
+    def apply(double: Double): String = {
+      if (double.isNaN || double.isInfinity) {
+        throw new IllegalArgumentException(s"Double value $double is not renderable to JSON.")
```
Comment (farmdawgnation) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: Can we change this to `Double value $double is not renderable to JSON when FailToRenderSpecialDoubleValues is used for rendering`

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala [hunk 6]
```diff
@@ -881,17 +881,74 @@ object JsonAST {
...
+        double.toString
+      }
+    }
+  }
+  /**
+   * A `DoubleRenderer` that throws an `IllegalArgumentException` when the
+   * special values `NaN`, `-Infinity`, and `Infinity` are encountered. Other
+   * doubles are rendered normally using `toString`.
+   */
+  case object FailToRenderSpecialValues extends DoubleRenderer {
```
Comment (farmdawgnation) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: I would prefer the name `FailToRenderSpecialDoubleValues` for clarity when talking about what the class does or when a curious explorer `println`s it.

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala [hunk 7]
```diff
@@ -881,17 +881,74 @@ object JsonAST {
...
+    def apply(double: Double): String = {
+      double.toString
+    }
+  }
+  /**
+   * A `DoubleRenderer` that renders special values `NaN`, `-Infinity`, and
+   * `Infinity` as `null`. Other doubles are rendered normally using
+   * `toString`.
+   */
+  case object RenderSpecialValuesAsNull extends DoubleRenderer {
```
Comment (farmdawgnation) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: Would prefer this to be named `RenderSpecialDoubleValuesAsNull` for the same reasons I mentioned elsewhere.

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala [hunk 8]
```diff
@@ -881,17 +881,74 @@ object JsonAST {
...
+  }
+  /**
+   * A `DoubleRenderer` that renders special values `NaN`, `-Infinity`, and
+   * `Infinity` as-is using `toString`. This is not strictly valid JSON, but
+   * JavaScript can eval it. Other double values are also rendered the same
+   * way.
+   *
+   * Usage is not recommended.
+   */
+  case object RenderSpecialValuesAsIs extends DoubleRenderer {
```
Comment (farmdawgnation) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: Would prefer this to be named `RenderSpecialDoubleValuesAsIs` for the same reasons mentioned elsewhere.

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala (hunk 1 above)
Comment (Shadowfiend) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: We already do, right? “This is not strictly valid JSON, but JavaScript can eval it”. Is there a rewording you'd prefer? Or is it the emphasis on Lift specifically not being able to reparse you're going for?

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala (hunk 6 above)
Comment (Shadowfiend) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: I hate this idea, but I think it's probably still a good idea, so I'll do it hehe.

File: core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala (hunk 2 above)
Comment (Shadowfiend) [inline core/json/src/test/scala/net/liftweb/json/JsonPrintingSpec.scala]: Whoops, yep. Will update these.

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala (hunk 5 above)
Comment (Shadowfiend) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: I'm loath to use the class name in a string. I can reword this to make it clearer that it's a choice, however.

**Comment by Shadowfiend** [comment]
Hit those requested changes.

File: core/json/src/main/scala/net/liftweb/json/JsonAST.scala (hunk 1 above)
Comment (farmdawgnation) [inline core/json/src/main/scala/net/liftweb/json/JsonAST.scala]: It's the emphasis on Lift specifically not being able to reparse it. Once could read this and interpret it as "This isn't strictly valid JSON even though we can write/parse it" which would not be true.

**Comment by farmdawgnation** [comment]
Your changes look good. Once we land in agreement on the comment we're discussing above, I'm good.
